### PR TITLE
Refactor key protection to be more C++ and less redundant.

### PR DESCRIPTION
**Pull request by ni4**

### CODE REVIEW

**Comment by ronaldtse** [comment]
Can we rerun the tests? #1505  is merged.

**Comment by ni4** [comment]
@ronaldtse Sure, I'll rebase.

**Review by ronaldtse — APPROVED** [review]
Thanks @ni4 !

**Comment by ronaldtse** [comment]
Ping @antonsviridenko for a quick check. Thanks!

File: src/lib/pgp-key.cpp [hunk 1]
```diff
@@ -1618,56 +1609,47 @@ pgp_key_t::add_protection(pgp_key_store_format_t             format,
...
                    const std::string &                new_password)
 {
     if (!is_secret()) {
         RNP_LOG("Warning: this is not a secret key");
         return false;
     }
-    if (!decrypted_seckey.material.secret) {
+    bool ownpkt = &decrypted == &pkt_;
+    if (!decrypted.material.secret) {
         RNP_LOG("Decrypted seckey must be provided");
```
Comment (antonsviridenko) [inline src/lib/pgp-key.cpp:1626]: maybe "secret key" instead of "seckey"?

**Comment by antonsviridenko** [comment]
Do we have some tests that check if secret key is actually protected? Like: get a pointer to the buffer with secret key material, copy contents, try to protect the key, then try to check if key contents are still available at previously stored address? It maybe bit complicated with everything moved to C++-way of things

File: src/lib/pgp-key.cpp (hunk 1 above)
Comment (ni4) [reply src/lib/pgp-key.cpp]: @antonsviridenko Thanks, fixed.

**Comment by ni4** [comment]
> Do we have some tests that check if secret key is actually protected? Like: get a pointer to the buffer with secret key material, copy contents, try to protect the key, then try to check if key contents are still available at previously stored address? It maybe bit complicated with everything moved to C++-way of things

We don't seem to have such tests, good idea, thanks. I'll update the PR.

**Comment by ni4** [comment]
@antonsviridenko Is it good to go once CI finished?
cc @ronaldtse

File: src/lib/pgp-key.cpp [hunk 2]
```diff
@@ -234,67 +234,59 @@ pgp_pk_alg_capabilities(pgp_pubkey_alg_t alg)
...
-        break;
-    case PGP_KEY_STORE_G10:
-        if (!g10_write_seckey(&memdst, seckey, password)) {
-            RNP_LOG("failed to write g10 seckey");
+    try {
+        switch (format) {
+        case PGP_KEY_STORE_GPG:
+        case PGP_KEY_STORE_KBX:
+            if (!write_sec_pgp(memdst, seckey, password)) {
+                RNP_LOG("failed to write seckey");
```
Comment (antonsviridenko) [inline src/lib/pgp-key.cpp:273]: same issue, "secret key"
probably I missed it last time :)

File: src/lib/pgp-key.cpp [hunk 3]
```diff
@@ -234,67 +234,59 @@ pgp_pk_alg_capabilities(pgp_pubkey_alg_t alg)
...
+                goto done;
+            }
+            break;
+        case PGP_KEY_STORE_G10:
+            if (!g10_write_seckey(&memdst, &seckey, password.c_str())) {
+                RNP_LOG("failed to write g10 seckey");
```
Comment (antonsviridenko) [inline src/lib/pgp-key.cpp:279]: "seckey"

**Review by antonsviridenko — CHANGES_REQUESTED** [review]

File: src/lib/pgp-key.cpp (hunk 3 above)
Comment (ni4) [reply src/lib/pgp-key.cpp:279]: Actually this message and previous one were not changed by this PR, but let's make things more understandable, will force-push shortly.

**Review by antonsviridenko — APPROVED** [review]

**Review by ronaldtse — APPROVED** [review]
Thanks @ni4 @antonsviridenko !

**Pull request merged**